### ISSUE TITLE
Distribute kaitai_struct_webide as Electron application and npm module

### ISSUE THREAD

**Issue by markbook2** [kaitai-io/kaitai_struct#154]
Distribute the `kaitai_struct_webide` as Electron application and [npm](https://www.npmjs.com/) module. it should allow us install the kaitai_struct_ide as npm package, something like that: `npm install kaitai` (alternative names: `npm install kaitai_struct` or `npm install kaitai_struct_ide`).

Here is an example of wrapping it with electron:
```sh
#!/usr/bin/env bash

set -euo pipefail

git clone --depth 1 https://github.com/kaitai-io/kaitai_struct_webide.git
cd kaitai_struct_webide
cat <<'OUT' > package.json
{
  "name": "kaitai_struct_webide",
  "description": "Web IDE for Kaitai Struct",
  "directories": {
    "doc": "docs"
  },
  "dependencies": {
    "typescript": "^2.2.1",
    "electron": "latest"
  },
  "main": "main.js",
  "devDependencies": {},
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "start-server": "python2 serve.py &",
    "start": "npm run-script start-server && electron ."
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/kaitai-io/kaitai_struct_webide.git"
  },
  "author": "",
  "license": "GPL-3.0",
  "bugs": {
    "url": "https://github.com/kaitai-io/kaitai_struct_webide/issues"
  },
  "homepage": "https://github.com/kaitai-io/kaitai_struct_webide#readme"
}
OUT
npm install
```
Create slightly changed kaitai_struct_webide/main.js from [electron quickstart](https://github.com/electron/electron-quick-start/blob/master/main.js):
```javascript
const electron = require('electron')
// Module to control application life.
const app = electron.app
// Module to create native browser window.
const BrowserWindow = electron.BrowserWindow

const path = require('path')
const url = require('url')

// Keep a global reference of the window object, if you don't, the window will
// be closed automatically when the JavaScript object is garbage collected.
let mainWindow

function createWindow () {
  // Create the browser window.
  mainWindow = new BrowserWindow({width: 800, height: 600})

  // and load the index.html of the app.
  mainWindow.loadURL('http://localhost:8000');

  // Open the DevTools.
  // mainWindow.webContents.openDevTools()

  // Emitted when the window is closed.
  mainWindow.on('closed', function () {
    // Dereference the window object, usually you would store windows
    // in an array if your app supports multi windows, this is the time
    // when you should delete the corresponding element.
    mainWindow = null
  })
}

// This method will be called when Electron has finished
// initialization and is ready to create browser windows.
// Some APIs can only be used after this event occurs.
app.on('ready', createWindow)

// Quit when all windows are closed.
app.on('window-all-closed', function () {
  // On OS X it is common for applications and their menu bar
  // to stay active until the user quits explicitly with Cmd + Q
  if (process.platform !== 'darwin') {
    app.quit()
  }
})

app.on('activate', function () {
  // On OS X it's common to re-create a window in the app when the
  // dock icon is clicked and there are no other windows open.
  if (mainWindow === null) {
    createWindow()
  }
})

// In this file you can include the rest of your app's specific main process
// code. You can also put them in separate files and require them here.
````
Then run with `npm start`.

It should work but it shows the `Your browser is not supported. Features may or may not working. Please use the latest Chrome or Firefox` message. It is shown in other browsers too (e.g. Firefox 52), so I assume we need to check compatibility.

**Comment by koczkatamas** [comment]
Currently Electron support is a low-priority issue for me: my concern is not about creating an app, but maintaining and supporting it.

I mentioned a few issues here: https://github.com/kaitai-io/kaitai_struct_webide/issues/20#issuecomment-292710273

Also I have security concerns too: https://twitter.com/bcrypt/status/852316922683076608

What would be the advantage of having a desktop app now? I see two things:
 - read/write files from filesystem: I want to solve this with a server-side component which can be run separately on your machine and you can use ide.kaitai.io to accessing your files
  - it's WIP actually: https://github.com/kaitai-io/kaitai_struct_webide/blob/master/serve_files.py
 - code signature: TLS protects your from ide.kaitai.io, even if we would create a desktop app and sign it, a malicious party could change to published pubkey and resign with his own cert

So currently I don't have the resources to maintain a desktop app in good conscience (keep it always up-to-date, make sure it won't cause any security issues, etc).

**Comment by markbook2** [comment]
It would be good to assign the low priority for this issue. Here are some thoughts about the subject:
1) Using Electron as any not controlled project can be not safe. So it would be good to have tools for exploring any software in a sandbox.
2) Beside the 1, it would provide the following:
- Single target browser (Chromium like)
- More users of Kaitai due publication through npmjs.com
- Removing dependency from Python and replacing it with js code
- It would be good to consider (if not done yet) of an ability unify the code and use Scala.js for everything
- Ability to expand the resulting npm package with another npm packages.

These items can consume many resources and time.

**Comment by koczkatamas** [comment]
I think I may have missing a point here...

I don't understand fully what you mean with your first point. Could you rephrase it?

> Single target browser (Chromium like)

You can get a similar feeling for any website of you use this Chrome feature: http://www.makeuseof.com/tag/use-chrome-create-windows-10-app-website/

Otherwise currently the only differences would be between using an Electron app and using ide.kaitai.io are the drawbacks: no updates, less security, more memory usage.

> Removing dependency from Python and replacing it with js code

Currently there is no python dependency. You have to host the html/css/js files by any means. `python -mSimpleHTTPServer` is the easiest way I know of after using ide.kaitai.io.

In the future for local file system access you would need python indeed, but python is installed by default on OSX and Linux. For Windows, I can create a simple .NET based server which will use much less resources than a bloated JS solution.

> It would be good to consider (if not done yet) of an ability unify the code and use Scala.js for everything

The compiler is compiled from Scala to JS and then bundled into the WebIDE. The whole package can be downloaded from here: https://github.com/kaitai-io/ide-kaitai-io.github.io. Then you just have to host it.

If you meant something else by this point, please rephrase it.

> More users of Kaitai due publication through npmjs.com
> Ability to expand the resulting npm package with another npm packages.

I am not familiar enough with the npm infrastructure to decide on this one. @GreyCat any experience, maybe?

**Comment by GreyCat** [comment]
> I am not familiar enough with the npm infrastructure to decide on this one. @GreyCat any experience, maybe?

I'm not a npm expert in any way, so I actually don't even understand what "Ability to expand the resulting npm package with another npm packages" is.

Technically, it should be a good idea to publish our JS runtime (and may be even the compiler) as npm package. Probably it would help people who want to integerate KS-based parsers (or ksc) into their software.

I don't think that we'll get much public visibility from npmjs.com. At the very least, maven repo + rubygems + python pip generate us so little traffic, so basically every little announce brings us more attention.

**Comment by markbook2** [comment]
> I don't understand fully what you mean with your first point. Could you rephrase it?

I mean that I don't have enough information about Electron's internals and propose to learn it better before using.

The goal of this issue is attempt to get closer to installation as one unit with minimal effort and getting working tool right after installation.

Another goal is modular architecture. Let's consider a hypothetical example: assume that user can install the `kaitai_struct_webide` with the `npm` command:
```
npm install kaitai_struct_webide
```
Then he can do:
```
npm start
```
and IDE UI in the browser (I call Electron browser here) should be shown. Then he can create own npm package say `my_webide` which will expand the `kaitai_struct_webide` and have dependency from it:
```json
{
  "name": "my_webide",
  "dependencies": {
    "kaitai_struct_webide": "latest"
  }
}
```
and publish it on npmjs.com too.

Of course we already have an ability of expansion (and modification) of the `kaitai_struct_webide` with the GitHub's `fork -> pull request` mechanisms. This issue suggests just another kind of expansion.

**Comment by markbook2** [comment]
I would like to clarity the following sentence in more detail:
> I mean that I don't have enough information about Electron's internals and propose to learn it better before using.

There is a forum topic called [Inox Browser - yet another Chromium Privacy Spin-off](https://bbs.archlinux.org/viewtopic.php?id=198763) - it is related to a fork of Chromium called Inox. Electron has common code with Chromium. So it is a possible that the following list copied from the forum's topic can be applied to Electron too:
```
based on patches:
    Disabled Google's Instant Extended API, this has the effect that the old "New Tab" is in place. (See disable-instant-extended-api.patch)
    DuckDuckGo as default search engine.            (See add-duckduckgo-seaarch-engine.patch)
    Disabled AutoFill data transmission                  (See disable-autofill-download-manager.patch)
    All Browser Extensions are now visible             (See modify-default-prefs.patch)
    Enabled user-modification for ALL extensions  (See disable-default-extensions.patch)
    Removed following default extensions:             (See disable-default-extensions.patch)
            -> Hotword (incl. Shared Module)
            -> Google Now
            -> Google Feedback
            -> Cloud Print
            -> Google Webstore
            -> Network Speech synthesis
            -> Google Hangout

    Disabled URLTracker data transmission. I know this class has a bad naming, but nevertheless it connects to Google. (See disable-google-url-tracker.patch)
    Disabled promo-notification fetching                     (See disable-notification-promo-fetch.patch)
    Disabled ipv6 probes to google servers                (See disable-google-ipv6-probes.patch)
    Disabled Google Cloud Messaging status check  (See disable-gcm-status-check.patch)

    Modified default settings:                                       (See modify-default-prefs.patch)
            -> DefaultCookiesSetting: CONTENT_SETTING_SESSION_ONLY
            -> EnableHyperLinkAuditing: false
            -> CloudPrintSubmitEnabled: false
            -> NetworkPredictionEnabled: false
            -> BackgroundModeEnabled: false
            -> BlockThirdPartyCookies: true
            -> AlternateErrorPagesEnabled: false
            -> SearchSuggestEnabled: false
            -> AutofillEnabled: false
            -> "Send feedback" checkbox if user triggers settings-reset: false
            -> BuiltInDnsClientEnabled: false
            -> SignInPromoUserSkipped: true
            -> SignInPromoShowOnFirstRunAllowed: false
            -> ShowAppsShortcutInBookmarkBar: false
            -> ShowBookmarkBar: true
            -> PromptForDownload: true
            -> SafeBrowsingEnabled: false
            -> EnableTranslate: false

based on build flags:
    Disabled google now:                enable_google_now=0
    Disabled WebRTC:                    enable_webrtc=0
    Disabled Remote service:          emable_remoting=0
    Disabled safe browsing:             safe_browsing_mode=0
    Disabled RLZ Identifier:              enable_rlz=0
    Disabled google hangouts:         enable_hangout_services_extension=0
    Disabled wifi bootstrapping:        enable_wifi_bootstrapping=0
    Disabled speech input:               enable_speech_input=0
    Disabled pre backups on sync:   enable_pre_sync_backup=0
    Disabled print preview:                enable_print_preview=0
    Disabled Chrome build:               google_chrome_build=0
```
I would  propose to check at least:
```
* Electron itself
* Used .so files
```
It is a possible that Electron don't have the mentioned above elements and don't  have other similar elements related to privacy.
